### PR TITLE
--[PBR Update 1 of 3] Map materials' extension/layer data to Drawables cache

### PR DESCRIPTION
**Pull request by jturner65**
## Motivation and Context
This PR is the first of 3 (or possibly 4) PRs to implement Habitat's new PBR materials extension support, [drawn from the original WIP found here](https://github.com/facebookresearch/habitat-sim/pull/2091).

In this PR the gltf extension data present in the Magnum material are loaded into each drawable's existing cache structure, to be more easily accessed each draw.  No changes to the shader are present in this PR.

## How Has This Been Tested
Existing C++ and python tests pass locally.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/gfx/PbrShader.cpp [hunk 1]
```diff
@@ -119,10 +119,8 @@ PbrShader::PbrShader(Flags originalFlags, unsigned int lightCount)
                      ? "#define NONE_ROUGHNESS_METALLIC_TEXTURE\n"
                      : "")
       .addSource(flags_ & Flag::NormalTexture ? "#define NORMAL_TEXTURE\n" : "")
-      .addSource(flags_ & Flag::NormalTextureScale
-                     ? "#define NORMAL_TEXTURE_SCALE\n"
-                     : "")
       .addSource(flags_ & Flag::ObjectId ? "#define OBJECT_ID\n" : "")
+      .addSource(flags_ & Flag::ClearCoatLayer ? "#define CLEAR_COAT\n" : "")
```
Comment (jturner65) [inline src/esp/gfx/PbrShader.cpp:123]: Oops one flag got past me as I cherry picked this. If this antagonizes anyone :grin: I'll get rid of it, but it's going to be added in the next PR anyway.

File: src/esp/gfx/GenericDrawable.cpp [hunk 2]
```diff
@@ -198,10 +198,8 @@ void GenericDrawable::draw(const Mn::Matrix4& transformationMatrix,
       // fragment shader
       .setObjectId(
           static_cast<RenderCamera&>(camera).useDrawableIds() ? drawableId_
-          : ((((flags_ & Mn::Shaders::PhongGL::Flag::InstancedObjectId) ==
-               Mn::Shaders::PhongGL::Flag::InstancedObjectId) ||
-              ((flags_ & Mn::Shaders::PhongGL::Flag::ObjectIdTexture)) ==
-                  Mn::Shaders::PhongGL::Flag::ObjectIdTexture))
+          : (((flags_ >= Mn::Shaders::PhongGL::Flag::InstancedObjectId) ||
```
Comment (eundersander) [inline src/esp/gfx/GenericDrawable.cpp:201]: Why do we compare flags with >=? Isn't flags basically a bitfield?

File: src/esp/gfx/PbrDrawable.cpp [hunk 3]
```diff
@@ -56,13 +58,17 @@ void PbrDrawable::setMaterialValuesInternal(
   matCache.metalness = tmpMaterialData.metalness();
   matCache.emissiveColor = tmpMaterialData.emissiveColor();
 
-  if (materialData_->hasAttribute("metallicTexturePointer") &&
-      materialData_->hasAttribute("roughnessTexturePointer")) {
+  const Cr::Containers::Optional<Mn::GL::Texture2D*> metallicTexturePtr =
```
Comment (eundersander) [inline src/esp/gfx/PbrDrawable.cpp:61]: This is pretty verbose. Could we use `auto` for the type of `metallicTexturePtr`? And/or a local, reusable lambda function?

File: src/esp/gfx/PbrDrawable.cpp [hunk 4]
```diff
@@ -117,31 +123,265 @@ void PbrDrawable::setMaterialValuesInternal(
...
     }
-    if (tmpMaterialData.normalTextureScale() != 1.0f) {
-      flags_ |= PbrShader::Flag::NormalTextureScale;
-      CORRADE_ASSERT(tmpMaterialData.normalTextureScale() > 0.0f,
-                     "PbrDrawable::PbrDrawable(): the normal texture scale "
-                     "must be positive.", );
-    }
+    // normal texture scale
+    matCache.normalTextureScale = tmpMaterialData.normalTextureScale();
+    CORRADE_ASSERT(tmpMaterialData.normalTextureScale() > 0.0f,
```
Comment (eundersander) [inline src/esp/gfx/PbrDrawable.cpp:136]: I realize you didn't add this assert, but anyway it would be nice to convert it to an ESP_CHECK. Doesn't feel like an assert.

File: src/esp/gfx/PbrShader.h [hunk 5]
```diff
@@ -172,41 +164,102 @@ class PbrShader : public Magnum::GL::AbstractShaderProgram {
...
+    InstancedObjectId = (1 << 11) | ObjectId,
+
+    /**
+     * Has ClearCoat layer.
+     */
+    ClearCoatLayer = 1 << 12,
+    /**
+     * Has ClearCoat Texture in ClearCoat layer
+     */
+    ClearCoatTexture = (1 << 13) | ClearCoatLayer,
```
Comment (eundersander) [inline src/esp/gfx/PbrShader.h:188]: I haven't seen this convention before so maybe a comment explaining. I guess you are basically grouping similar enum values to have a shared bit.

File: src/esp/gfx/PbrShader.h [hunk 6]
```diff
@@ -131,18 +131,10 @@ class PbrShader : public Magnum::GL::AbstractShaderProgram {
      */
     NormalTexture = 1 << 6,
 
-    /**
```
Comment (eundersander) [inline src/esp/gfx/PbrShader.h:134]: Just curious, why was this removed? (My guess: your new version of the shader will always use a scale, so you didn't need this flag anymore.)

File: src/esp/gfx/PbrShader.h (hunk 5 above)
Comment (eundersander) [reply src/esp/gfx/PbrShader.h:188]: Also, I see we're using these with `Corrade::EnumSet` and from [here](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html) I see "The only requirement for the enum type is that all the values must be binary exclusive." @mosra thoughts?

**Review by eundersander — COMMENTED** [review]
I left some comments! I'll let others weigh in on the enum usage concerns before approving.

File: src/esp/gfx/GenericDrawable.cpp (hunk 2 above)
Comment (jturner65) [reply src/esp/gfx/GenericDrawable.cpp:201]: that's how magnum compares multi-bit flags

File: src/esp/gfx/PbrShader.h (hunk 6 above)
Comment (jturner65) [reply src/esp/gfx/PbrShader.h:134]: Yep. Talked to mosra about it and it's sufficiently cheap to always enable.

File: src/esp/gfx/PbrDrawable.cpp (hunk 4 above)
Comment (mosra) [reply src/esp/gfx/PbrDrawable.cpp:136]: Won't negative scales simply invert the normal texture, i.e. a weird but valid use case? I don't think this need an assertion at all.

File: src/esp/gfx/PbrShader.h (hunk 5 above)
Comment (mosra) [reply src/esp/gfx/PbrShader.h:188]: I'll remove that line from the docs. Having a particular flag a superset of another is a very common use case, so the "requirement" makes no sense nowadays.

File: src/esp/gfx/GenericDrawable.cpp (hunk 2 above)
Comment (mosra) [reply src/esp/gfx/GenericDrawable.cpp:201]: [Docs for reference](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html#aeb270daa9a8afb3a80e57ef2cb2f8314), `>=` is chosen because it looks close to a ⊇, a symbol for a superset.

File: src/esp/gfx/PbrDrawable.cpp (hunk 3 above)
Comment (mosra) [reply src/esp/gfx/PbrDrawable.cpp:61]: I don't like `auto` but I think it's fine to use it in this case -- the type is almost clear from the findAttribute call below.

File: src/esp/gfx/PbrDrawable.cpp (hunk 4 above)
Comment (jturner65) [reply src/esp/gfx/PbrDrawable.cpp:136]: how about abs(matCache.normalTextureScale) > 0?

Comment (jturner65) [reply src/esp/gfx/PbrDrawable.cpp]: oh, it won't bug out when I try to query it in the shader?  cool

**Review by Skylion007 — APPROVED** [review]

File: src/esp/gfx/PbrDrawable.cpp (hunk 4 above)
Comment (mosra) [reply src/esp/gfx/PbrDrawable.cpp]: Why would it? You simply get a zero value, so as long as you don't divide anything with it, you'll get a valid result.

File: src/esp/gfx/GenericDrawable.cpp (hunk 2 above)
Comment (eundersander) [reply src/esp/gfx/GenericDrawable.cpp:201]: Ok so, what the heck do these lines mean? :) Is it testing whether `flags_` includes `InstanceObjectId`?

Comment (mosra) [reply src/esp/gfx/GenericDrawable.cpp:201]: > Is it testing whether `flags_` includes `InstanceObjectId`?

Yes. It's a shorthand for `(a & other) == other`, as the docs say. It does exactly what it did before -- checking if one set of bits is a superset of the other set -- just in a more concise way.
